File: backend/studi0trace/engines/vexel/reuse.py
"""Stage 10b: repeated shapes as `<use>`.

A dot grid, a row of bullets, a tiled pattern: the same shape painted many
times. Each copy is traced on its own and the copies agree to a few hundredths
of a pixel, so the file says so — the first copy's geometry goes into `<defs>`
once, with no fill of its own, and every copy is a `<use>` that names it,
carries its own translation and its own fill. Painter's order is kept: each
`<use>` stands exactly where the copy's element would have.

Two shapes are copies when they are the same kind of primitive with the same
dimensions to USE_TOL, or two paths whose outlines, one moved onto the other
by the difference of their bounding boxes, lie within USE_TOL of each other
both ways (mean nearest distance).
"""
from __future__ import annotations

import numpy as np
from scipy.spatial import cKDTree

from studi0trace.engines.vexel.curves import (
    CircArc,
    Circle,
    Cubic,
    Ellipse,
    Line,
    PathShape,
    Rect,
    RoundedRect,
    Shape,
    _f,
    arc_points,
    path_d,
)

USE_TOL = 0.10        # px; copies agree to this
USE_MIN_POINTS = 8    # a path this small is not worth a reference


def _sample(shape: PathShape, per_segment: int = 12) -> np.ndarray:
    t = np.linspace(0.0, 1.0, per_segment)[:, None]
    out = []
    for contour in shape.contours:
        for seg in contour:
            if isinstance(seg, Line):
                out.append(seg.p0 * (1 - t) + seg.p1 * t)
            elif isinstance(seg, CircArc):
                out.append(arc_points(seg, per_segment))
            else:
                out.append((1 - t) ** 3 * seg.p0 + 3 * (1 - t) ** 2 * t * seg.c1 + 3 * (1 - t) * t ** 2 * seg.c2 + t ** 3 * seg.p1)
    return np.vstack(out) if out else np.zeros((0, 2))


def _signature(shape: Shape) -> tuple:
    if isinstance(shape, PathShape):
        return ("path", tuple(tuple(type(s).__name__ for s in c) for c in shape.contours))
    return (type(shape).__name__,)


def _origin(shape: Shape, pts: np.ndarray | None) -> np.ndarray:
    if isinstance(shape, (Circle, Ellipse)):
        return np.array([shape.cx, shape.cy])
    if isinstance(shape, (Rect, RoundedRect)):
        return np.array([shape.x, shape.y])
    return pts.min(axis=0)


def _same(a: Shape, b: Shape, pa: np.ndarray | None, pb: np.ndarray | None) -> np.ndarray | None:
    """The translation carrying `a` onto `b` when they are copies, else None."""
    if _signature(a) != _signature(b):
        return None
    if isinstance(a, Circle):
        return np.array([b.cx - a.cx, b.cy - a.cy]) if abs(a.r - b.r) <= USE_TOL else None
    if isinstance(a, Ellipse):
        ok = abs(a.rx - b.rx) <= USE_TOL and abs(a.ry - b.ry) <= USE_TOL and abs(a.angle_deg - b.angle_deg) <= 0.5
        return np.array([b.cx - a.cx, b.cy - a.cy]) if ok else None
    if isinstance(a, Rect):
        ok = abs(a.w - b.w) <= USE_TOL and abs(a.h - b.h) <= USE_TOL
        return np.array([b.x - a.x, b.y - a.y]) if ok else None
    if isinstance(a, RoundedRect):
        ok = abs(a.w - b.w) <= USE_TOL and abs(a.h - b.h) <= USE_TOL and abs(a.rx - b.rx) <= USE_TOL
        return np.array([b.x - a.x, b.y - a.y]) if ok else None
    if len(pa) < USE_MIN_POINTS or len(pb) != len(pa):
        return None
    shift = pb.min(axis=0) - pa.min(axis=0)
    moved = pa + shift
    if float(cKDTree(pb).query(moved)[0].mean()) > USE_TOL or float(cKDTree(moved).query(pb)[0].mean()) > USE_TOL:
        return None
    return shift


def _translated(shape: Shape, shift: np.ndarray) -> Shape:
    """The shape moved by −shift is the definition; we move the *first* copy to
    the origin of the definition, which is itself, so this is only used to
    write the definition without a fill."""
    return shape


def _geometry(shape: Shape, sid: str, precision: int) -> str:
    """The definition: geometry only, no paint."""
    from studi0trace.engines.vexel.curves import shape_svg

    return shape_svg(shape, f'id="{sid}"', precision)


def emit(pending: list[tuple[Shape, str]], precision: int, first_id: int = 1) -> tuple[list[str], list[str]]:
    """`pending` is (shape, attrs) in paint order. Returns (defs, elements):
    every element is either the shape's own markup or a `<use>` of a definition
    written once into `defs`."""
    samples = [(_sample(s) if isinstance(s, PathShape) else None) for s, _a in pending]
    group_of: list[int | None] = [None] * len(pending)
    shifts: list[np.ndarray | None] = [None] * len(pending)
    groups: list[list[int]] = []
    for i in range(len(pending)):
        if group_of[i] is not None:
            continue
        members = [i]
        for j in range(i + 1, len(pending)):
            if group_of[j] is not None:
                continue
            shift = _same(pending[i][0], pending[j][0], samples[i], samples[j])
            if shift is not None:
                group_of[j] = len(groups)
                shifts[j] = shift
                members.append(j)
        if len(members) > 1:
            group_of[i] = len(groups)
            shifts[i] = np.zeros(2)
            groups.append(members)
    defs: list[str] = []
    elements: list[str] = []
    ids: dict[int, str] = {}
    for g, members in enumerate(groups):
        sid = f"u{first_id + g}"
        ids[g] = sid
        defs.append(_geometry(pending[members[0]][0], sid, precision))
    for i, (shape, attrs) in enumerate(pending):
        g = group_of[i]
        if g is None:
            from studi0trace.engines.vexel.curves import shape_svg

            elements.append(shape_svg(shape, attrs, precision))
            continue
        dx, dy = shifts[i]
        pos = "" if abs(dx) < 1e-9 and abs(dy) < 1e-9 else f' x="{_f(float(dx), precision)}" y="{_f(float(dy), precision)}"'
        elements.append(f'<use href="#{ids[g]}"{pos} {attrs}/>')
    return defs, elements

File: backend/vexel-rs/src/reuse.rs
//! Stage 10b: repeated shapes as `<use>`. Mirrors `reuse.py`.

use crate::curves::{fmt, shape_svg, Segment, Shape, P};
use crate::symmetry::Grid;

pub const USE_TOL: f64 = 0.10;
pub const USE_MIN_POINTS: usize = 8;

pub(crate) fn arc_centre(p0: P, p1: P, r: f64, large: bool, sweep: bool) -> P {
    let mid = [0.5 * (p0[0] + p1[0]), 0.5 * (p0[1] + p1[1])];
    let d = [p1[0] - p0[0], p1[1] - p0[1]];
    let half = 0.5 * (d[0] * d[0] + d[1] * d[1]).sqrt();
    if half < 1e-12 {
        return mid;
    }
    let r = r.max(half);
    let h = (r * r - half * half).max(0.0).sqrt();
    let n = [-d[1] / (2.0 * half), d[0] / (2.0 * half)];
    if sweep != large { [mid[0] + n[0] * h, mid[1] + n[1] * h] } else { [mid[0] - n[0] * h, mid[1] - n[1] * h] }
}

/// `n` points along an arc, ends included. Mirrors the Python `arc_points`.
pub fn arc_points(p0: P, p1: P, r: f64, large: bool, sweep: bool, n: usize) -> Vec<P> {
    let c = arc_centre(p0, p1, r, large, sweep);
    let a0 = (p0[1] - c[1]).atan2(p0[0] - c[0]);
    let a1 = (p1[1] - c[1]).atan2(p1[0] - c[0]);
    let tau = 2.0 * std::f64::consts::PI;
    let span = if sweep { (a1 - a0).rem_euclid(tau) } else { -((a0 - a1).rem_euclid(tau)) };
    let rr = ((p0[0] - c[0]).powi(2) + (p0[1] - c[1]).powi(2)).sqrt();
    (0..n)
        .map(|k| {
            let t = a0 + span * k as f64 / (n - 1) as f64;
            [c[0] + rr * t.cos(), c[1] + rr * t.sin()]
        })
        .collect()
}

fn sample(contours: &[Vec<Segment>]) -> Vec<P> {
    let per = 12usize;
    let mut out = Vec::new();
    for contour in contours {
        for seg in contour {
            match seg {
                Segment::Line { p0, p1 } => {
                    for k in 0..per {
                        let t = k as f64 / (per - 1) as f64;
                        out.push([p0[0] * (1.0 - t) + p1[0] * t, p0[1] * (1.0 - t) + p1[1] * t]);
                    }
                }
                Segment::Arc { p0, p1, r, large, sweep } => out.extend(arc_points(*p0, *p1, *r, *large, *sweep, per)),
                Segment::Cubic { p0, c1, c2, p1 } => {
                    for k in 0..per {
                        let t = k as f64 / (per - 1) as f64;
                        let mt = 1.0 - t;
                        let (a, b, c, d) = (mt * mt * mt, 3.0 * mt * mt * t, 3.0 * mt * t * t, t * t * t);
                        out.push([a * p0[0] + b * c1[0] + c * c2[0] + d * p1[0], a * p0[1] + b * c1[1] + c * c2[1] + d * p1[1]]);
                    }
                }
            }
        }
    }
    out
}

fn signature(shape: &Shape) -> String {
    match shape {
        Shape::Path { contours } => {
            let mut s = String::from("path:");
            for c in contours {
                for seg in c {
                    s.push(match seg {
                        Segment::Line { .. } => 'L',
                        Segment::Cubic { .. } => 'C',
                        Segment::Arc { .. } => 'A',
                    });
                }
                s.push('|');
            }
            s
        }
        Shape::Circle { .. } => "Circle".into(),
        Shape::Ellipse { .. } => "Ellipse".into(),
        Shape::Rect { .. } => "Rect".into(),
        Shape::RoundedRect { .. } => "RoundedRect".into(),
    }
}

fn bbox_min(pts: &[P]) -> P {
    let mut m = [f64::INFINITY, f64::INFINITY];
    for p in pts {
        m[0] = m[0].min(p[0]);
        m[1] = m[1].min(p[1]);
    }
    m
}

fn chamfer(from: &[P], to: &[P]) -> f64 {
    let grid = Grid::new(to);
    from.iter().map(|p| grid.nearest(*p).0).sum::<f64>() / from.len() as f64
}

/// The translation carrying `a` onto `b` when they are copies. See the Python `_same`.
fn same(a: &Shape, b: &Shape, pa: Option<&Vec<P>>, pb: Option<&Vec<P>>) -> Option<P> {
    if signature(a) != signature(b) {
        return None;
    }
    match (a, b) {
        (Shape::Circle { cx, cy, r }, Shape::Circle { cx: bx, cy: by, r: br }) => {
            if (r - br).abs() <= USE_TOL { Some([bx - cx, by - cy]) } else { None }
        }
        (Shape::Ellipse { cx, cy, rx, ry, angle_deg }, Shape::Ellipse { cx: bx, cy: by, rx: brx, ry: bry, angle_deg: bang }) => {
            if (rx - brx).abs() <= USE_TOL && (ry - bry).abs() <= USE_TOL && (angle_deg - bang).abs() <= 0.5 { Some([bx - cx, by - cy]) } else { None }
        }
        (Shape::Rect { x, y, w, h }, Shape::Rect { x: bx, y: by, w: bw, h: bh }) => {
            if (w - bw).abs() <= USE_TOL && (h - bh).abs() <= USE_TOL { Some([bx - x, by - y]) } else { None }
        }
        (Shape::RoundedRect { x, y, w, h, rx }, Shape::RoundedRect { x: bx, y: by, w: bw, h: bh, rx: brx }) => {
            if (w - bw).abs() <= USE_TOL && (h - bh).abs() <= USE_TOL && (rx - brx).abs() <= USE_TOL { Some([bx - x, by - y]) } else { None }
        }
        (Shape::Path { .. }, Shape::Path { .. }) => {
            let (pa, pb) = (pa?, pb?);
            if pa.len() < USE_MIN_POINTS || pb.len() != pa.len() {
                return None;
            }
            let (ma, mb) = (bbox_min(pa), bbox_min(pb));
            let shift = [mb[0] - ma[0], mb[1] - ma[1]];
            let moved: Vec<P> = pa.iter().map(|p| [p[0] + shift[0], p[1] + shift[1]]).collect();
            if chamfer(&moved, pb) > USE_TOL || chamfer(pb, &moved) > USE_TOL {
                return None;
            }
            Some(shift)
        }
        _ => None,
    }
}

/// `pending` is (shape, attrs) in paint order. Returns (defs, elements). See the Python `emit`.
pub fn emit(pending: &[(Shape, String)], precision: usize, first_id: usize) -> (Vec<String>, Vec<String>) {
    let samples: Vec<Option<Vec<P>>> = pending
        .iter()
        .map(|(s, _)| if let Shape::Path { contours } = s { Some(sample(contours)) } else { None })
        .collect();
    let n = pending.len();
    let mut group_of: Vec<Option<usize>> = vec![None; n];
    let mut shifts: Vec<Option<P>> = vec![None; n];
    let mut groups: Vec<Vec<usize>> = Vec::new();
    for i in 0..n {
        if group_of[i].is_some() {
            continue;
        }
        let mut members = vec![i];
        for j in i + 1..n {
            if group_of[j].is_some() {
                continue;
            }
            if let Some(shift) = same(&pending[i].0, &pending[j].0, samples[i].as_ref(), samples[j].as_ref()) {
                group_of[j] = Some(groups.len());
                shifts[j] = Some(shift);
                members.push(j);
            }
        }
        if members.len() > 1 {
            group_of[i] = Some(groups.len());
            shifts[i] = Some([0.0, 0.0]);
            groups.push(members);
        }
    }
    let mut defs = Vec::new();
    let mut ids: Vec<String> = Vec::new();
    for (g, members) in groups.iter().enumerate() {
        let sid = format!("u{}", first_id + g);
        defs.push(shape_svg(&pending[members[0]].0, &format!("id=\"{}\"", sid), precision));
        ids.push(sid);
    }
    let mut elements = Vec::with_capacity(n);
    for (i, (shape, attrs)) in pending.iter().enumerate() {
        match group_of[i] {
            None => elements.push(shape_svg(shape, attrs, precision)),
            Some(g) => {
                let [dx, dy] = shifts[i].unwrap();
                let pos = if dx.abs() < 1e-9 && dy.abs() < 1e-9 {
                    String::new()
                } else {
                    format!(" x=\"{}\" y=\"{}\"", fmt(dx, precision), fmt(dy, precision))
                };
                elements.push(format!("<use href=\"#{}\"{} {}/>", ids[g], pos, attrs));
            }
        }
    }
    (defs, elements)
}
